metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/utils/RdfRetriever.java

```diff
@@ -19,25 +20,89 @@
    * 
    * @param resourceId The remote entity to retrieve (resource IDs are in fact URIs)
-   * @return The string representation of the remote entity
+   * @param possibleSuffixes The suffixes we will try to append to the entity to form the remote
+   *        address.
+   * @return The original entity containing a string representation of the remote entity. This
+   *         method does not return null, but the returned entity may contain a null XML value.
    */
-  public String retrieve(String resourceId) {
-    if (resourceId != null) {
-      try {
-        URLConnection urlConnection = new URL(resourceId).openConnection();
-        urlConnection.setRequestProperty("accept", "application/rdf+xml");
-        InputStream inputStream = urlConnection.getInputStream();
-        StringWriter writer = new StringWriter();
-        IOUtils.copy(inputStream, writer, StandardCharsets.UTF_8);
-        return writer.toString();
-      } catch (IOException e) {
-        LOGGER.error("Failed to retrieve: {} with message: {}", resourceId, e.getMessage());
+  public String retrieve(String resourceId, Set<String> possibleSuffixes) {
+
+    // Get the entity from the own store
+    OriginalEntity originalEntity = entityDao.get(resourceId);
+
+    // If we can't find it, get it from the remote source.
+    if (originalEntity == null) {
+      originalEntity = retrieveFromSource(resourceId, possibleSuffixes);
+      entityDao.save(originalEntity);
+    }
+
+    // Done.
+    return originalEntity.getXml();
+  }
+
+  private OriginalEntity retrieveFromSource(String resourceId, Set<String> possibleSuffixes) {
+
+    // Sanity check for null values.
+    if (resourceId == null) {
+      throw new IllegalArgumentException("Parameter resourceId cannot be null.");
+    }
+
+    // Try to retrieve the entity for the different suffixes, stopping when we succeed.
+    String content = null;
+    for (String suffix : possibleSuffixes) {
+      content = retrieveFromSource(resourceId, suffix);
+      if (content != null) {
+        break;
       }
     }
-    return "";
+
+    // Compile the entity, regardless of whether we have any content.
+    OriginalEntity originalEntity = new OriginalEntity();
+    originalEntity.setURI(resourceId);
+    originalEntity.setXml(content);
+    return originalEntity;
+  }
+
+  private static String retrieveFromSource(String resourceId, String suffix) {
+
+    // Make the connection and retrieve the result.
+    String result = null;
+    String resultContentType = null;
+    try {
+      URLConnection urlConnection = new URL(resourceId + suffix).openConnection();
+      urlConnection.setRequestProperty("accept", "application/rdf+xml");
+      result = IOUtils.toString(urlConnection.getInputStream(), StandardCharsets.UTF_8);
+      resultContentType = urlConnection.getContentType();
+    } catch (IOException e) {
+      LOGGER.error("Failed to retrieve: {} with message: {}", resourceId, e.getMessage());
```

Comment:
It doesn't, but when it does, it probably does a lot (e.g. for a whole batch of records if the user made a mistake in the vocabulary configuration). I don't want to flood the logs in this case. But I suppose that adding a debug statement here would be OK?

Reply:
I saw some notification, but the comment has disappeared, and yes debug level seems to be more appropriate indeed.

Reply:
Yes, I removed it because I was considering an alternative solution. I'm not anymore: the comment basically said that I would add a debug statement with the full stacktrace.